In [49]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

In [50]:
# reading datasets
bb_usage = pd.read_csv("data/user_profile/BB_Usage_Categorical.csv")
voice_onnet_incoming_usage = pd.read_csv("data/user_profile/VOICE_ONNET_INCOMING_Usage_Categorical.csv")
voice_onnet_outgoing_usage = pd.read_csv("data/user_profile/VOICE_ONNET_OUTGOING_Usage_Categorical.csv")
voice_offnet_incoming_usage = pd.read_csv("data/user_profile/VOICE_OFFNET_INCOMING_Usage_Categorical.csv")
voice_offnet_outgoing_usage = pd.read_csv("data/user_profile/VOICE_OFFNET_OUTGOING_Usage_Categorical.csv")
voice_IDD_incoming_usage = pd.read_csv("data/user_profile/VOICE_IDD_Incoming_Usage_Categorical.csv")
voice_IDD_outgoing_usage = pd.read_csv("data/user_profile/VOICE_IDD_Outgoing_Usage_Categorical.csv")
peo_tv_usage = pd.read_csv("data/user_profile/PEO_TV_Usage_Categorical.csv")

# product details
phone_num_products = pd.read_csv("data/user_profile/Products_Associated_to_Phone_No.csv")


In [51]:
# locations
bb_location = pd.read_csv("data/user_profile/BB_USAGE_Locations_of_Users.csv")
voice_onnet_incoming_location = pd.read_csv("data/user_profile/VOICE_ONNET_INCOMING_Locations_of_Users.csv")
voice_onnet_outgoing_location = pd.read_csv("data/user_profile/VOICE_ONNET_OUTGOING_Locations_of_Users.csv")
voice_offnet_incoming_location = pd.read_csv("data/user_profile/VOICE_OFFNET_INCOMING_Locations_of_Users.csv")
voice_offnet_outgoing_location = pd.read_csv("data/user_profile/VOICE_OFFNET_OUTGOING_Locations_of_Users.csv")
voice_IDD_incoming_location = pd.read_csv("data/user_profile/VOICE_IDD_INCOMING_Locations_of_Users.csv")
peo_tv_location = pd.read_csv("data/user_profile/PEO_TV_Locations_of_Users.csv")

# Data perapration

In [52]:
bb_usage.rename(columns={'Unnamed: 0':'event_source.hash'}, inplace="True")
bb_usage = bb_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled']]
bb_usage.drop_duplicates("event_source.hash", inplace=True)
bb_usage.drop([0,1], axis=0, inplace=True)
bb_usage

event_source.hash Usage_Rating_Categorical  \
2     00567a4a3c474aca1ce5cd6570648932                   MEDIUM   
3     00724ae2470df1ed0fe919800d02517d                      LOW   
4     00a8c3e3baed0557c31ab9b68a086ed2                      LOW   
5     00cc5a820586ce291c0fa4c2a99a4816                      LOW   
6     00d1f690ec4af27702febbd0db6942c2                      LOW   
...                                ...                      ...   
1303  feca6cdb74a4a95dec4fe88f3fb2f938                      LOW   
1304  ff1093d5ef61bf46f8c88e9b7614e311                      LOW   
1305  ff45fdfb4b6dac191aba17a281181d39                   MEDIUM   
1306  ffb2ec238eba276f65157693ed329c7f                   MEDIUM   
1307  ffd697d66cc01c8897ce3e37d693673a                      LOW   

      Usage_Rating_Scaled  
2                0.619792  
3                0.031250  
4                0.175347  
5                0.072917  
6                0.128472  
...                   ...  
1303             0.041667  
1304             0.326389  
1305             0.625000  
1306             0.567708  
1307             0.013889  

[1306 rows x 3 columns]

In [53]:
voice_onnet_incoming_usage = voice_onnet_incoming_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_onnet_incoming_usage.drop_duplicates("event_source.hash", inplace=True)
voice_onnet_incoming_usage

event_source.hash Usage_Rating_Categorical  \
0     00221451f705ebe26051158bb14f567a                      LOW   
1     002fb47f60400713c854b0f69ff78c0a                      LOW   
2     00567a4a3c474aca1ce5cd6570648932                      LOW   
3     006be321aad541ff1d6c32a43dd7cdde                   MEDIUM   
4     00724ae2470df1ed0fe919800d02517d                      LOW   
...                                ...                      ...   
3231  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                   MEDIUM   
3232  ffb63ff99cf1354dff1da862f379da80                   MEDIUM   
3233  ffd697d66cc01c8897ce3e37d693673a                      LOW   
3234  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   
3235  ffee845c0caa320b55e31bb156827d16                   MEDIUM   

      Usage_Rating_Scaled Outlier  
0                0.017514      NO  
1                0.056319      NO  
2                0.001374      NO  
3                0.108688      NO  
4                0.002747      NO  
...                   ...     ...  
3231             0.080529      NO  
3232             0.173935      NO  
3233             0.005495      NO  
3234             0.203984      NO  
3235             0.112466      NO  

[3236 rows x 4 columns]

In [54]:
voice_onnet_outgoing_usage = voice_onnet_outgoing_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_onnet_outgoing_usage.drop_duplicates("event_source.hash", inplace=True)
voice_onnet_outgoing_usage

event_source.hash Usage_Rating_Categorical  \
0     00221451f705ebe26051158bb14f567a                      LOW   
1     002fb47f60400713c854b0f69ff78c0a                      LOW   
2     00567a4a3c474aca1ce5cd6570648932                      LOW   
3     00724ae2470df1ed0fe919800d02517d                      LOW   
4     007400db8afa996e788d23986739f3db                      LOW   
...                                ...                      ...   
3302  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                      LOW   
3303  ffb63ff99cf1354dff1da862f379da80                   MEDIUM   
3304  ffd697d66cc01c8897ce3e37d693673a                      LOW   
3305  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   
3306  ffee845c0caa320b55e31bb156827d16                      LOW   

      Usage_Rating_Scaled Outlier  
0                0.009321      NO  
1                0.051339      NO  
2                0.002404      NO  
3                0.003140      NO  
4                0.025755      NO  
...                   ...     ...  
3302             0.053964      NO  
3303             0.227924      NO  
3304             0.002943      NO  
3305             0.213108      NO  
3306             0.043122      NO  

[3307 rows x 4 columns]

In [55]:
voice_offnet_incoming_usage = voice_offnet_incoming_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_offnet_incoming_usage.drop_duplicates("event_source.hash", inplace=True)
voice_offnet_incoming_usage

event_source.hash Usage_Rating_Categorical  \
0     00221451f705ebe26051158bb14f567a                      LOW   
1     002fb47f60400713c854b0f69ff78c0a                      LOW   
2     00567a4a3c474aca1ce5cd6570648932                      LOW   
3     006be321aad541ff1d6c32a43dd7cdde                   MEDIUM   
4     00724ae2470df1ed0fe919800d02517d                      LOW   
...                                ...                      ...   
3324  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                      LOW   
3325  ffb63ff99cf1354dff1da862f379da80                   MEDIUM   
3326  ffd697d66cc01c8897ce3e37d693673a                      LOW   
3327  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   
3328  ffee845c0caa320b55e31bb156827d16                      LOW   

      Usage_Rating_Scaled Outlier  
0                0.085835      NO  
1                0.079989      NO  
2                0.002404      NO  
3                0.227801      NO  
4                0.010817      NO  
...                   ...     ...  
3324             0.011332      NO  
3325             0.147142      NO  
3326             0.012019      NO  
3327             0.135744      NO  
3328             0.035878      NO  

[3329 rows x 4 columns]

In [56]:
voice_offnet_outgoing_usage = voice_offnet_outgoing_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_offnet_outgoing_usage.drop_duplicates("event_source.hash", inplace=True)
voice_offnet_outgoing_usage

event_source.hash Usage_Rating_Categorical  \
0     00221451f705ebe26051158bb14f567a                      LOW   
1     002fb47f60400713c854b0f69ff78c0a                      LOW   
2     00567a4a3c474aca1ce5cd6570648932                      LOW   
3     00724ae2470df1ed0fe919800d02517d                      LOW   
4     007400db8afa996e788d23986739f3db                      LOW   
...                                ...                      ...   
3186  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                      LOW   
3187  ffb63ff99cf1354dff1da862f379da80                     HIGH   
3188  ffd697d66cc01c8897ce3e37d693673a                      LOW   
3189  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   
3190  ffee845c0caa320b55e31bb156827d16                      LOW   

      Usage_Rating_Scaled Outlier  
0                0.048813      NO  
1                0.053424      NO  
2                0.003606      NO  
3                0.004808      NO  
4                0.039197      NO  
...                   ...     ...  
3186             0.018029      NO  
3187             0.334012      NO  
3188             0.001202      NO  
3189             0.196281      NO  
3190             0.021635      NO  

[3191 rows x 4 columns]

In [57]:
voice_IDD_incoming_usage = voice_IDD_incoming_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_IDD_incoming_usage.drop_duplicates("event_source.hash", inplace=True)
voice_IDD_incoming_usage

event_source.hash Usage_Rating_Categorical  \
0     002fb47f60400713c854b0f69ff78c0a                      LOW   
1     007400db8afa996e788d23986739f3db                      LOW   
2     009f305bf85a71012c0e343233e611ba                     HIGH   
3     011adeeaef79aeac706a2e7fc225e8f8                      LOW   
4     013736757d2ddad83cdfaa418921605c                      LOW   
...                                ...                      ...   
1192  fefee9a8aed82517d62c14ea4793bf0a                      LOW   
1193  ff445f580f7525bd9f1455d6b09704cd                      LOW   
1194  ff7e0f4edb7b39b683b9134af37fc4d3                     HIGH   
1195  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                      LOW   
1196  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   

      Usage_Rating_Scaled Outlier  
0                0.014423      NO  
1                0.043269      NO  
2                0.679487     YES  
3                0.032051      NO  
4                0.006410      NO  
...                   ...     ...  
1192             0.052885      NO  
1193             0.028846      NO  
1194             0.673077     YES  
1195             0.006410      NO  
1196             0.110577     YES  

[1197 rows x 4 columns]

In [58]:
voice_IDD_outgoing_usage = voice_IDD_outgoing_usage[['event_source.hash','Usage_Rating_Categorical','Usage_Rating_Scaled','Outlier']]
voice_IDD_outgoing_usage.drop_duplicates("event_source.hash", inplace=True)
voice_IDD_outgoing_usage

event_source.hash Usage_Rating_Categorical  \
0    007400db8afa996e788d23986739f3db                      LOW   
1    009f305bf85a71012c0e343233e611ba                      LOW   
2    00cc5a820586ce291c0fa4c2a99a4816                      LOW   
3    011adeeaef79aeac706a2e7fc225e8f8                      LOW   
4    0201ee9b60de662cdd16c885f178feea                      LOW   
..                                ...                      ...   
512  fe20b17a9573e4b7e765cc3098728f3d                      LOW   
513  fe4e907a21bb03842eb88d0c0d633ca8                     HIGH   
514  ff445f580f7525bd9f1455d6b09704cd                      LOW   
515  ff7e0f4edb7b39b683b9134af37fc4d3                     HIGH   
516  ffd96f173d6cd683bf3384bdc2f99713                   MEDIUM   

     Usage_Rating_Scaled Outlier  
0               0.004808      NO  
1               0.014423      NO  
2               0.032051      NO  
3               0.006410      NO  
4               0.030449      NO  
..                   ...     ...  
512             0.057692      NO  
513             0.375000     YES  
514             0.004808      NO  
515             0.826923     YES  
516             0.272436     YES  

[517 rows x 4 columns]

In [59]:
peo_tv_usage = peo_tv_usage[['Phone_Number.hash','Peo_TV_Package','Usage_Rating_Categorical','Usage_Rating_Scaled']]
peo_tv_usage.drop_duplicates("Phone_Number.hash", inplace=True)
peo_tv_usage

Phone_Number.hash   Peo_TV_Package  \
0     46960b77dad950ddae76dd53bd7848cd      PEO_UTHAYAM   
1     71d08664b609dcaddf02d6ffc2532aa7       PEO_SILVER   
2     f98bd7d6c96317fc2656bc43441df8bb       PEO_SILVER   
3     a41d8c2af7a4d478417f2368ee852296      PEO_UTHAYAM   
4     09e7342066a13ca589d4a58ea0c685ae       PEO_SILVER   
...                                ...              ...   
1423  3ec6441eb5df6ef9ba2bdd63f919ce8b       PEO_SILVER   
1424  013736757d2ddad83cdfaa418921605c       PEO_SILVER   
1425  5fc30724801d125e0048b486655723f6       PEO_SILVER   
1426  e034fe9b7cb567273a07edbad9946fde  PEO_SILVER_PLUS   
1427  2ab8436cc34b4f16de3c452300f975c3       PEO_SILVER   

     Usage_Rating_Categorical  Usage_Rating_Scaled  
0                        HIGH              0.93750  
1                      MEDIUM              0.40625  
2                        HIGH              0.75000  
3                          NO              0.00000  
4                        HIGH              0.90625  
...                       ...                  ...  
1423                     HIGH              1.00000  
1424                       NO              0.00000  
1425                     HIGH              0.75000  
1426                     HIGH              0.78125  
1427                      LOW              0.15625  

[1428 rows x 4 columns]

In [60]:
# user location 
print(bb_location.shape)
print(voice_onnet_incoming_location.shape)
print(voice_onnet_outgoing_location.shape)
print(voice_offnet_incoming_location.shape)
print(voice_offnet_outgoing_location.shape)
print(voice_IDD_incoming_location.shape)
print(peo_tv_location.shape)

(1306, 3)
(3236, 3)
(3307, 3)
(3329, 3)
(3191, 3)
(1197, 3)
(1428, 4)


In [61]:
# location from voice onnet outgoing
# justification: this has the max no of subscribers
voice_onnet_outgoing_location = voice_onnet_outgoing_location[['event_source.hash','Location_Identifier']]
voice_onnet_outgoing_location.drop_duplicates("event_source.hash", inplace=True)
voice_onnet_outgoing_location

event_source.hash              Location_Identifier
0     d7b523c1eba8e755b5c8a149f6562348  MSAG5200-ISLIDH-KGW-NODE6194491
1     e66925d5277dac781824d054f75998d7  MSAG5200-ISLIDH-KGW-NODE6194491
2     f7508647677b63cfa0f63fe548f3236b  MSAG5200-ISLIDH-KGW-NODE6194491
3     4c01c7a8616e48c2b2dc8c56bb09b605    MSAG5200-ISLPC-AGT-NODE480031
4     27e3e9cde496878b297d35ee9ef8406a         C300MNW-MEE-NODE18887751
...                                ...                              ...
3302  59ce55487af1cbf6855b860ab5455019    MSAG5200-ISLVA-NKM-NODE481961
3303  15d5e1896fc1afc13074efcd398edf8e    MSAG5200-ISLND-PHG-NODE500241
3304  ef1e2374e34fa4f73b6b4ae1e640d318    MSAG5200-ISLND-PHG-NODE500241
3305  8bce192f1a4cf21bcd65958a32f537ed        MSAG5200PH-AKG-NODE431631
3306  6d219e6c2e086d6959f4d90cad121009        MSAG5200PH-AKG-NODE431631

[3307 rows x 2 columns]

In [62]:
phone_num_products

Unnamed: 0                 Phone_Number.hash  \
0             0  009f305bf85a71012c0e343233e611ba   
1             1  01cdd813d9a8bea5d70455a917c91e63   
2             2  028c5d688b270307aea69a8dcb74bee2   
3             3  05cfe8f4323299542b61cfdb60232095   
4             4  06c40958738d7ff752abdd8973327e84   
..          ...                               ...   
436         436  fc49108a316204817f42189f9f6e6a70   
437         437  fc77db5a9efeccfaace51889f68e3f22   
438         438  fd62e39a7c82e9e9a066aee4307fdf87   
439         439  feebdc83f80ba16208b54c2902471410   
440         440  ffb9cb2e838925de5a75764fd59c6ea1   

                                          PRODUCT_NAME  Product_Count  \
0    ['V_Abbreviated Dialing', 'V_Call Transfer Thr...             13   
1    ['V_Call Waiting', 'V_Call Forwarding-Immediat...              8   
2                      ['E_TSTV', 'E_Video on Demand']              2   
3    ['E_PeoTV Initiation', 'E_TSTV', 'E_Video on D...              3   
4    ['V_DQ Contact Details', 'V_IDD', 'AB_Copper A...              5   
..                                                 ...            ...   
436  ['V_Referral offer Discount', 'AB_Referral off...              2   
437                    ['E_Video on Demand', 'E_TSTV']              2   
438         ['V_Call Forwarding', 'V_Hotline Service']              2   
439                    ['E_Video on Demand', 'E_TSTV']              2   
440  ['AB_Megaline', 'V_SLT CLI', 'V_IDD', 'AB_Mega...              5   

                                          Product_Dict  
0    {'009f305bf85a71012c0e343233e611ba': ['V_Abbre...  
1    {'01cdd813d9a8bea5d70455a917c91e63': ['V_Call ...  
2    {'028c5d688b270307aea69a8dcb74bee2': ['E_TSTV'...  
3    {'05cfe8f4323299542b61cfdb60232095': ['E_PeoTV...  
4    {'06c40958738d7ff752abdd8973327e84': ['V_DQ Co...  
..                                                 ...  
436  {'fc49108a316204817f42189f9f6e6a70': ['V_Refer...  
437  {'fc77db5a9efeccfaace51889f68e3f22': ['E_Video...  
438  {'fd62e39a7c82e9e9a066aee4307fdf87': ['V_Call ...  
439  {'feebdc83f80ba16208b54c2902471410': ['E_Video...  
440  {'ffb9cb2e838925de5a75764fd59c6ea1': ['AB_Mega...  

[441 rows x 5 columns]

# User Profile : Usage

In [65]:
usage_df = voice_onnet_outgoing_usage.merge(voice_onnet_incoming_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical_x':'ONNET_OUTGOING_Categorical', 'Usage_Rating_Categorical_y':'ONNET_INCOMING_Categorical',
                         'Usage_Rating_Scaled_x':'ONNET_OUTGOING_Scaled','Usage_Rating_Scaled_y':'ONNET_INCOMING_Scaled',
                         'Outlier_x':'ONNET_OUTGOING_Outlier','Outlier_y':'ONNET_INCOMING_Outlier'}, inplace="True")

usage_df = usage_df.merge(voice_offnet_outgoing_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'OFFNET_OUTGOING_Categorical','Usage_Rating_Scaled':'OFFNET_OUTGOING_Scaled', 'Outlier':'OFFNET_OUTGOING_Outlier'}, inplace="True")

usage_df = usage_df.merge(voice_offnet_incoming_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'OFFNET_INCOMING_Categorical','Usage_Rating_Scaled':'OFFNET_INCOMING_Scaled','Outlier':'OFFNET_INCOMING_Outlier'}, inplace="True")

usage_df = usage_df.merge(voice_IDD_incoming_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'IDD_INCOMING_Categorical','Usage_Rating_Scaled':'IDD_INCOMING_Scaled','Outlier':'IDD_INCOMING_Outlier'}, inplace="True")

usage_df = usage_df.merge(voice_IDD_outgoing_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'IDD_OUTGOING_Categorical','Usage_Rating_Scaled':'IDD_OUTGOING_Scaled','Outlier':'IDD_OUTGOING_Outlier'}, inplace="True")

usage_df = usage_df.merge(peo_tv_usage,how='left', left_on='event_source.hash', right_on='Phone_Number.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'PEO_TV_Categorical','Usage_Rating_Scaled':'PEO_TV_Scaled'}, inplace="True")
usage_df.drop('Phone_Number.hash', axis=1, inplace=True)

# merege BB usage
usage_df = usage_df.merge(bb_usage,how='left', on='event_source.hash')
usage_df.rename(columns={'Usage_Rating_Categorical':'BB_Categorical','Usage_Rating_Scaled':'BB_Scaled'}, inplace="True")

# merge user location identifier
usage_df = usage_df.merge(voice_onnet_outgoing_location,how='left', on='event_source.hash')

# merge Product packages map
usage_df = usage_df.merge(phone_num_products[['Phone_Number.hash','PRODUCT_NAME','Product_Count']],how='left', left_on='event_source.hash', right_on='Phone_Number.hash')
usage_df.rename(columns={'PRODUCT_NAME':'Other_Product_Packages','Product_Count':'Other_Product_Packages_Count'}, inplace="True")
usage_df.drop('Phone_Number.hash', axis=1, inplace=True)

cat_cols = ['ONNET_OUTGOING_Categorical', 'ONNET_INCOMING_Categorical', 'OFFNET_OUTGOING_Categorical', 'OFFNET_INCOMING_Categorical',
            'IDD_INCOMING_Categorical','IDD_OUTGOING_Categorical','Peo_TV_Package', 'PEO_TV_Categorical','BB_Categorical','Other_Product_Packages'
           ,'IDD_OUTGOING_Outlier','OFFNET_OUTGOING_Outlier','ONNET_OUTGOING_Outlier','ONNET_INCOMING_Outlier','ONNET_INCOMING_Outlier'
           ,'IDD_INCOMING_Outlier']

scaled_cols = ['ONNET_OUTGOING_Scaled', 'ONNET_INCOMING_Scaled', 'OFFNET_OUTGOING_Scaled', 'OFFNET_INCOMING_Scaled','IDD_INCOMING_Scaled'
        ,'IDD_OUTGOING_Scaled','PEO_TV_Scaled','BB_Scaled','Other_Product_Packages_Count']

usage_df[cat_cols] = usage_df[cat_cols].fillna("NO")

usage_df[scaled_cols] = usage_df[scaled_cols].fillna(0)
usage_df

event_source.hash ONNET_OUTGOING_Categorical  \
0     00221451f705ebe26051158bb14f567a                        LOW   
1     002fb47f60400713c854b0f69ff78c0a                        LOW   
2     00567a4a3c474aca1ce5cd6570648932                        LOW   
3     00724ae2470df1ed0fe919800d02517d                        LOW   
4     007400db8afa996e788d23986739f3db                        LOW   
...                                ...                        ...   
3302  ffb4fa6778bc5b3d9c858f5f4fa1e0c2                        LOW   
3303  ffb63ff99cf1354dff1da862f379da80                     MEDIUM   
3304  ffd697d66cc01c8897ce3e37d693673a                        LOW   
3305  ffd96f173d6cd683bf3384bdc2f99713                     MEDIUM   
3306  ffee845c0caa320b55e31bb156827d16                        LOW   

      ONNET_OUTGOING_Scaled ONNET_OUTGOING_Outlier ONNET_INCOMING_Categorical  \
0                  0.009321                     NO                        LOW   
1                  0.051339                     NO                        LOW   
2                  0.002404                     NO                        LOW   
3                  0.003140                     NO                        LOW   
4                  0.025755                     NO                        LOW   
...                     ...                    ...                        ...   
3302               0.053964                     NO                     MEDIUM   
3303               0.227924                     NO                     MEDIUM   
3304               0.002943                     NO                        LOW   
3305               0.213108                     NO                     MEDIUM   
3306               0.043122                     NO                     MEDIUM   

      ONNET_INCOMING_Scaled ONNET_INCOMING_Outlier  \
0                  0.017514                     NO   
1                  0.056319                     NO   
2                  0.001374                     NO   
3                  0.002747                     NO   
4                  0.017857                     NO   
...                     ...                    ...   
3302               0.080529                     NO   
3303               0.173935                     NO   
3304               0.005495                     NO   
3305               0.203984                     NO   
3306               0.112466                     NO   

     OFFNET_OUTGOING_Categorical  OFFNET_OUTGOING_Scaled  \
0                            LOW                0.048813   
1                            LOW                0.053424   
2                            LOW                0.003606   
3                            LOW                0.004808   
4                            LOW                0.039197   
...                          ...                     ...   
3302                         LOW                0.018029   
3303                        HIGH                0.334012   
3304                         LOW                0.001202   
3305                      MEDIUM                0.196281   
3306                         LOW                0.021635   

     OFFNET_OUTGOING_Outlier  ... IDD_OUTGOING_Scaled  IDD_OUTGOING_Outlier  \
0                         NO  ...            0.000000                    NO   
1                         NO  ...            0.000000                    NO   
2                         NO  ...            0.000000                    NO   
3                         NO  ...            0.000000                    NO   
4                         NO  ...            0.004808                    NO   
...                      ...  ...                 ...                   ...   
3302                      NO  ...            0.000000                    NO   
3303                      NO  ...            0.000000                    NO   
3304                      NO  ...            0.000000                    NO   
3305                      NO  ...            0.272436      

In [66]:
 len(usage_df[usage_df['Other_Product_Packages'] == 'NO'])

2907

In [67]:
3307- 2907

400

In [68]:
usage_with_other_packages = usage_df[usage_df['Other_Product_Packages'] != 'NO']
usage_with_other_packages

event_source.hash ONNET_OUTGOING_Categorical  \
7     009f305bf85a71012c0e343233e611ba                     MEDIUM   
24    01cdd813d9a8bea5d70455a917c91e63                        LOW   
71    05cfe8f4323299542b61cfdb60232095                        LOW   
79    06c40958738d7ff752abdd8973327e84                     MEDIUM   
100   08c38286f1a189b638e60f800de4cf98                        LOW   
...                                ...                        ...   
3239  f987efdcbfcdc0f3f78d4ee6ed37b88a                        LOW   
3261  fc49108a316204817f42189f9f6e6a70                        LOW   
3263  fc77db5a9efeccfaace51889f68e3f22                     MEDIUM   
3275  fd62e39a7c82e9e9a066aee4307fdf87                        LOW   
3293  feebdc83f80ba16208b54c2902471410                        LOW   

      ONNET_OUTGOING_Scaled ONNET_OUTGOING_Outlier ONNET_INCOMING_Categorical  \
7                  0.249951                     NO                       HIGH   
24                 0.031324                     NO                        LOW   
71                 0.022248                     NO                        LOW   
79                 0.193755                     NO                     MEDIUM   
100                0.001202                     NO                         NO   
...                     ...                    ...                        ...   
3239               0.006917                     NO                        LOW   
3261               0.046041                     NO                        LOW   
3263               0.088525                     NO                        LOW   
3275               0.058109                     NO                        LOW   
3293               0.007947                     NO                        LOW   

      ONNET_INCOMING_Scaled ONNET_INCOMING_Outlier  \
7                  0.593407                    YES   
24                 0.002747                     NO   
71                 0.009272                     NO   
79                 0.172047                     NO   
100                0.000000                     NO   
...                     ...                    ...   
3239               0.008929                     NO   
3261               0.017170                     NO   
3263               0.045330                     NO   
3275               0.039148                     NO   
3293               0.034684                     NO   

     OFFNET_OUTGOING_Categorical  OFFNET_OUTGOING_Scaled  \
7                           HIGH                0.575549   
24                           LOW                0.019599   
71                           LOW                0.068779   
79                        MEDIUM                0.103635   
100                          LOW                0.002404   
...                          ...                     ...   
3239                         LOW                0.010817   
3261                      MEDIUM                0.086342   
3263                         LOW                0.035224   
3275                         LOW                0.040031   
3293                         LOW                0.008413   

     OFFNET_OUTGOING_Outlier  ... IDD_OUTGOING_Scaled  IDD_OUTGOING_Outlier  \
7                        YES  ...            0.014423                    NO   
24                        NO  ...            0.000000                    NO   
71                        NO  ...            0.000000                    NO   
79                        NO  ...            0.000000                    NO   
100                       NO  ...            0.000000                    NO   
...                      ...  ...                 ...                   ...   
3239                      NO  ...            0.000000                    NO   
3261                      NO  ...            0.000000                    NO   
3263                      NO  ...            0.000000                    NO   
3275                      NO  ...            0.000000      

## Note:
- product pacakges merged using the available phone number(as **PRODUCT LABEL**) has only 400 matches. 
- Reason: Even "Prodduct state Changes df" has **3000+** , when grouped by **Account No** there are only **441 unque phone numbers**(event sources) associated with services. 

In [69]:
usage_df.to_csv("data/user_profile/prepared/User_Profile.csv")

## ToDo
- filter out free products
- need more product pacakges data( for remaining event sources)
- catagorize product packages according to **VOICE/ BB** (which are now in **other_product_packages** col
- enrish user profile with more features